contracts/extension/plugin/PluginData.sol

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "../interface/plugin/IPlugin.sol";
+import "../../lib/TWStringSet.sol";
+
+library PluginDataStorage {
+    bytes32 public constant PLUGIN_DATA_STORAGE_POSITION = keccak256("plugin.datastore.storage");
+
+    struct Data {
+        TWStringSet.Set pluginNames;
+        mapping(string => IPlugin.Plugin) plugins;
+        mapping(bytes4 => IPlugin.PluginMetadata) pluginMetadata;
+    }
+
+    function pluginDataStorage() internal pure returns (Data storage pluginData) {
+        bytes32 position = PLUGIN_DATA_STORAGE_POSITION;
+        assembly {
+            pluginData.slot := position
+        }
+    }
+}
+
+contract PluginData is IPlugin {
```

Comment:
so the logic to add/remove plugins from a Router is a plugin itself? :D

Reply:
The naming could be more specific though. How about `PluginHolder`, `PluginModifer`, `PluginState` or something along those lines?

Reply:
No this won't be a plugin itself 😅 Just a contract with common logic shared by all three `Router` `PluginMap` and `PluginRegistry`.

I can name it `PluginState` that makes sense.